### ISSUE TITLE
Error message doesn't show when password reset token has expired

### ISSUE THREAD

**Issue by avishkar58** [deep-learning-indaba/Baobab#710]
If you try to reset a password with an expired token, after clicking submit, nothing happens and no error is shown.

**Comment by avishkar58** [comment]
Hey @greeffjean - here's an easy one to get you started. Some more information to reproduce this locally: 
- Create a user account on the app running on localhost:5000 (look in your terminal for the verification email)
- Request a password reset and note the URL in the email in the terminal
- Update the date on the password reset table in the database (you can connect to it on port 5432 using psql / pgAdmin) to something in the past
- Browse to the reset URL and enter a new password
- Click submit
- Nothing happens in the UI (although the response from the API is indeed that the password reset token has expired)

**Comment by greeffjean** [comment]
Great Ill get cracking as soon as I get the project up and running!

**Comment by greeffjean** [comment]
Hi Avishkar.

Having some trouble please assist:

Create a user account on the app running on localhost:5000 (look in your
terminal for the verification email)
According to the App Readme.MD, the app only runs on localhost:8080 and I
am having trouble doing migrations:

Starting baobab-develop_db_1      ... error
Starting baobab-develop_storage_1 ... done

ERROR: for baobab-develop_db_1  Cannot start service db: Ports are not
available: listen tcp 0.0.0.0:5432: bind: address already in use

ERROR: for db  Cannot start service db: Ports are not available: listen tcp
0.0.0.0:5432: bind: address already in use
ERROR: Encountered errors while bringing up the project.

If it is so that I am to sign up on localhost:8080 then another issue is
the title selector in "http://localhost:8080/createAccount", there are no
options present here and my request will not submit because this field is
empty.

Please advise

**Comment by greeffjean** [comment]
Please supply a little more guidance here: I am not familiar with this task

Update the date on the password reset table in the database (you can
connect to it on port 5432 using psql / pgAdmin) to something in the past

**Comment by avishkar58** [comment]
Hi Jean,

Apologies, I meant localhost:8080 which is the front-end. localhost:5000 is the REST API.

The errors suggest that the database is already running on that port. Does the baobab-develop_db container start up correctly when you run the entire app with `docker-compose up`? Did you run the migrations while the app was already running? Can you try shut down the application and run the migrations again? Which operating system are you on by the way? 

The reason the title options aren't available in the selector is because they are retrieved from the rest API, which in turn gets them from the database, but since you haven't successfully run the migrations, there is no data in the database. Once you've run the migrations it should work

Regarding your final question, are you familiar with relational databases and SQL? The purpose of this is just to hack the data in your local database to make the reset token expired, rather than waiting the 2 weeks for it to expire on its own. To do this, download pgAdmin (any OS) or Postico (which I prefer, but is Mac only). Ensure that you've run the migrations and that the docker containers are running and connect to the database on localhost, port 5432 in pgAdmin/Postico. The username and password are both `docker`. Open the `docker` database and you'll see a list of the tables, double click on the password_reset table then you'll see the rows of data in the table and should be able to directly edit the date on the relevant row. I suppose you could fix this issue without doing this hack, but I don't know how else you'll test it.

**Comment by greeffjean** [comment]
Hi Avishkar

Im using a Mac, the baobab-develop_db container does not appear to be
starting up please see here:

Starting baobab-develop_db_1 ...
Starting baobab-develop_db_1      ... error
baobab-develop_webappintegration_1 is up-to-date

Starting baobab-develop_test-ci_1 ... done

ERROR: for db  Cannot start service db: Ports are not available: listen tcp
0.0.0.0:5432: bind: address already in use
ERROR: Encountered errors while bringing up the project.

I have run the migrations once the app was running, I think the issue is
here in the docker-compose up step?

Regards
Jean

**Comment by greeffjean** [comment]
Hi, did have a look at your code as well, I think I may have found the
issue in the UI, there was one too many </div> tags that would be breaking
the code. I do need to test it out though locally to be sure.

Waiting for you to assist me with getting set up.

Regards
Jean

**Comment by avishkar58** [comment]
Hi Jean,

Are you running any other services on port 5432 ? You can check with
netstat or lsof:

netstat -vanp tcp | grep 5432

Or

sudo lsof -i tcp:5432

**Comment by greeffjean** [comment]
Yes I believe I am, I have another project set up on my local machine here is the response: 

tcp4       0      0  *.5432                 *.*                    LISTEN      131072 131072    147      0 0x0000 0x00000006
tcp6       0      0  *.5432                 *.*                    LISTEN      131072 131072    147      0 0x0000 0x00000006
justines-iMac:baobab-develop justinecrewebrown$ sudo lsof -i tcp:5432
Password:
COMMAND  PID     USER   FD   TYPE             DEVICE SIZE/OFF NODE NAME
postgres 147 postgres    4u  IPv6 0x2f8f11be0db6336f      0t0  TCP *:postgresql (LISTEN)
postgres 147 postgres    5u  IPv4 0x2f8f11be0c4763df      0t0  TCP *:postgresql (LISTEN)

**Comment by greeffjean** [comment]
Solution?  Im not too clued up on these types of set ups

**Comment by avishkar58** [comment]
Ok you can change the port that Postgres binds to in [docker-compose.yml, line number 8](https://github.com/deep-learning-indaba/Baobab/blob/309837a6de8744d2c94340534e11a6ef9eb3e61b/docker-compose.yml#L8)

Then you can change the port the application connects to in [api/config.py line 13](https://github.com/deep-learning-indaba/Baobab/blob/309837a6de8744d2c94340534e11a6ef9eb3e61b/api/config.py#L13) - since it's just for your local copy, just replace the `os.getenv('DB_PORT_5432_TCP_ADDR')` with whatever port you chose

**Comment by KaleabTessera** [comment]
@greeffjean Try running `docker stop $(docker ps -aq)` and then re-running the above. 

It looks like you have containers running on certain ports that haven't been stopped. That command stops all containers. 

Have a look at https://www.docker.com/sites/default/files/d8/2019-09/docker-cheat-sheet.pdf and other resources to get a hang of docker.

**Comment by KaleabTessera** [comment]
@avishkar58 's way would also work, but it might be easier to just restart the containers, then change the port.

**Comment by greeffjean** [comment]
Up and running !

**Comment by avishkar58** [comment]
Great! :)

**Comment by greeffjean** [comment]
Just about to go buy a ssd drive lol, things a bit slow my end.

**Comment by greeffjean** [comment]
Not getting into the database please see screenshots : 
I have ensures that the migrations and that the docker containers are running and connect to the database on localhost.
![Screenshot 2020-07-15 at 15 59 22](https://user-images.githubusercontent.com/44994396/87554376-5aeb3b80-c6b4-11ea-80ec-d6533302d86c.png)

**Comment by greeffjean** [comment]
![Screenshot 2020-07-15 at 16 01 11](https://user-images.githubusercontent.com/44994396/87554444-6dfe0b80-c6b4-11ea-9fef-8a2947dc2fda.png)

**Comment by greeffjean** [comment]
Docker-compose.yml : 

redis:
  image: redis
  ports:
    - 6379:6379
db:
  image: orchardup/postgresql
  ports:
    - 1234:1234

**Comment by greeffjean** [comment]
api/config.py : 

SQLALCHEMY_DATABASE_URI = os.getenv('DATABASE_URL',
                                    'postgres://docker:docker@{0}/docker'.format(os.getenv('DB_PORT_1234_TCP_ADDR')))

**Comment by avishkar58** [comment]
Can you try 127.0.0.1 as the host?

**Comment by avishkar58** [comment]
@KaleabTessera Any ideas?

**Comment by greeffjean** [comment]
That did not work Im afraid

**Comment by greeffjean** [comment]
If its any help localhost:5000 in the browser :  status code 400

Bad Request
Could not resolve organisation

**Comment by avishkar58** [comment]
That's fine, the API expects the HTTP origin header to be set, which is why you get that error if you try to navigate directly to it in the browser. You can try a get request with Postman and set the Origin header to localhost if you want to interact with the API.

Can you confirm the database is actually running and listening for connections?
`sudo lsof -i tcp:5432` on my machine gives:  (replace 5432 with your port number)

```
COMMAND     PID     USER   FD   TYPE             DEVICE SIZE/OFF NODE NAME
Postico   15616 avishkar   19u  IPv6 0xe504f7dd018a4317      0t0  TCP localhost:59139->localhost:postgresql (CLOSE_WAIT)
com.docke 15810 avishkar   51u  IPv6 0xe504f7dd018a05d7      0t0  TCP *:postgresql (LISTEN)
com.docke 15810 avishkar   57u  IPv6 0xe504f7dd018a1837      0t0  TCP localhost:postgresql->localhost:59139 (FIN_WAIT_2)
```

**Comment by greeffjean** [comment]
Appers to be working 

`COMMAND   PID              USER   FD   TYPE             DEVICE SIZE/OFF NODE NAME
com.docke 705 justinecrewebrown   54u  IPv6 0x2f8f11be0db6210f      0t0  TCP *:search-agent (LISTEN)

`

**Comment by greeffjean** [comment]
Appers to be working

`COMMAND PID USER FD TYPE DEVICE SIZE/OFF NODE NAME
com.docke 705 justinecrewebrown 54u IPv6 0x2f8f11be0db6210f 0t0 TCP *:search-agent (LISTEN)

`

**Comment by greeffjean** [comment]
Sorry hit the closed button by mistake

**Comment by KaleabTessera** [comment]
@greeffjean  I think it is an issue related to your port configuration. 

If you need to run it on a diff port, you can't change both ports, one of the ports is the internal db port for docker, while the other one is external port used by your pc. 

For example,
`5432:5432` - internal port used by docker is 5432, external is 5432. 

I think  you need to change your ports to 
```db:
  image: orchardup/postgresql
  ports:
       5432:1234 (It might 1234:5432, I can't remember the order). 
```
That should push traffic from internal docker container port 5432 to your 1234 port.

**Comment by greeffjean** [comment]
Morning guys. I fixed the bug and tested my solution, 

Im going to push to this branch [iss710] so you can review this morning, just writing a test at the moment. 

Im assuming I am contributing in the correct way.. ?

**Comment by avishkar58** [comment]
Hi Jean,

That's great! Yes please push to your branch and send a pull request to
merge into develop for review.

**Comment by avishkar58** [comment]
Fixed in PR #713
